### ISSUE TITLE
Connection API accepts SDP for receivers lacking ts-refclk or mediaclk media attributes

### ISSUE THREAD

**Issue by billt-hlit** [sony/nmos-cpp#76]
As an example, the following PATCH request was accepted

`uuid=191941ea-84e5-52ae-81ee-ccf59bba9f1b; curl -v -H Content-Type: application/json -H Accept: application/json -X PATCH  http://d13:3215/x-nmos/connection/v1.0/single/receivers/191941ea-84e5-52ae-81ee-ccf59bba9f1b/staged  -d '{"master_enable": false, "transport_file": { "data": "v=0\r\no=- 1234 2 IN IP4 0.0.0.0\r\ns=!\r\nt=0 0\r\nm=audio 20111 RTP/AVP 97\r\ni=Second PCM audio pair, 24-bit samples\r\nc=IN IP4 228.164.214.111/32\r\na=source-filter:incl IN IP4 228.164.214.111 10.10.164.213\r\na=rtpmap:97 L24/48000/4\r\na=fmtp:97 channel-order=SMPTE2110.(ST)\r\na=ptime:1\r\na=mediaclk:direct=0\r\n", "type": "application/sdp"},"activation": {"mode": "activate_immediate"}}'`

For reference, SMPTE 2110-10 §8 requires these attributes.

**Comment by billt-hlit** [comment]
Relatedly and belatedly, the contents of the ts-refclk entry don't seem to get parsed very carefully, either. The following all get through:

```
a=ts-refclk:ntp=2368.8.8.8
a=ts-refclk:ntp=;28.8.8.8
a=ts-refclk:ptp=IEEE1588-2008:08-00-11-ff-fe-21-db-34:127:42
a=ts-refclk:ptp=IEEE1588-2008:408-00-11-ff-fe-21-db-34:127
a=ts-refclk:ptp=IEEE1588-2008:08-00-11-ff-fe-21-db-34:337
```

That said, I don't know how particular the parser can be. The first one, in particular, pushes up against the old "what is a host name?" question.

**Comment by garethsb-sony** [comment]
OK, reviewing `ts-refclk` and `mediaclk` first before getting on to the wider issue...

[ST 2110-10:2017](https://ieeexplore.ieee.org/document/8165974) Section 8.1 says:
> All stream descriptions shall have a media-level `mediaclk` attribute as per IETF RFC 7273 section 5.2 The `direct` reference should be used.

To be clear, [RFC 7273 Section 5.2](https://tools.ietf.org/html/rfc7273#section-5.2) defines how to signal a "direct-referenced media clock" but doesn't make it mandatory, that's all ST 2110-10. RFC 7273 also defines how "signalling included at session level provides default parameters for all RTP sessions and sources in the session description. More specific signalling included at the media level overrides session-level signalling." ST 2110-10 is overruling that, which seems somewhat heavy-handed.

ST 2110-10:2017 Section 8.2 says:
> All stream descriptions shall have a media-level `ts-refclk` attribute as specified in IETF RFC 7273 section 4.

And as above, [RFC 7273 Section 4](https://tools.ietf.org/html/rfc7273#section-4) defines "timestamp reference clock source signalling" but it's ST 2110-10 that's making it required at the media level.

Although NMOS is designed to support RTP streams that aren't ST 2110 (as well as other transports completely), it seems reasonable for `nmos::get_session_description_sdp_parameters` and `nmos::validate_sdp_parameters` etc. (declared in [nmos/sdp_utils.h](https://github.com/sony/nmos-cpp/blob/master/Development/nmos/sdp_utils.h)) to be focused on ST 2110.

However, since we know implementations exist that don't signal `ts-refclk` and `mediaclk` as required by ST 2110-10, what should we do?

It would be easy enough to reject them - e.g. for `ts-refclk`, a small change in [`nmos::get_session_description_sdp_parameters`](https://github.com/sony/nmos-cpp/blob/master/Development/nmos/sdp_utils.cpp#L773-L791) to `throw details::sdp_processing_error("missing timestamp reference clock source")` if it's not present.

However, I can imagine implementations wanting to "do their best" with such incoming SDP files?

**Comment by garethsb-sony** [comment]
To recap how the [Connection API](https://github.com/AMWA-TV/nmos-device-connection-management/blob/v1.0/APIs/ConnectionAPI.raml) is intended to reject `PATCH /staged` requests that it doesn't like, it's a 400 response if the JSON is invalid (syntax or schema) or doesn't meet the `/constraints`, or a 500 "if the transport file could not be parsed, or there is some failure caused by activation which was not caused by a violation or schema or constraints."

The case that we're talking about of staging an SDP file that _can_ be parsed, but isn't valid according to ST 2110 falls in a grey area therefore. But probably a 500 response to the `PATCH /staged` request is expected, even if the request didn't specify activation.

In nmos-cpp, we don't actually have the ability to allow user code to validate a Connection API request until activation, and that's arguably too late. :disappointed:

One potential solution is to use more state-transitions in the Connection API as @billt-hlit originally proposed, storing the merged request in e.g. "staging", next to "staged" and "active". I am coincidentally wondering about using an "auto" sibling to allow simpler but greater flexibility than the current `resolve_auto` approach.

**Comment by garethsb-sony** [comment]
A better solution to validating `PATCH /staged` requests before they are applied, let alone before activation, may be to add `std::function` callbacks (or hooks as @jwpwh has called them in #77, etc.) to `nmos::make_connection_api` (the Connection API factory function). This makes sense for things which do not need to be asynchronous, such as parsing and validating the SDP file (also cf. #38, #39, #74) and possibly for `resolve_auto` at the point of activation.

**Comment by billt-hlit** [comment]
I like your callback idea. The callback could even enforce more semantic constraints (like "this receiver only supports HD") that might be particular to the hardware.

**Comment by garethsb-sony** [comment]
I think all of the above can be dealt with once PR #81 is merged.

Handle "staging an SDP file that _can_ be parsed, but isn't valid according to ST 2110" or has "things that the actual node implementation doesn't like", or "enforce more semantic constraints (like "this receiver only supports HD")" by providing a custom `nmos::transport_file_parser` callback and/or a custom `nmos::details::connection_resource_patch_validator` callback.

**Comment by garethsb-sony** [comment]
Closed by #81.
